pyomo/util/latex_printer.py

```diff
@@ -0,0 +1,1336 @@
+
+from pyomo.common import DeveloperError
+
+_CONSTANT = ExprType.CONSTANT
+_MONOMIAL = ExprType.MONOMIAL
+_GENERAL = ExprType.GENERAL
+
+from pyomo.common.errors import InfeasibleConstraintException
+
+from pyomo.common.dependencies import numpy, numpy_available
+
+if numpy_available:
+    import numpy as np
+
+
+def decoder(num, base):
+    # Needed in the general case, but not as implemented
+    # if isinstance(base, float):
+    #     if not base.is_integer():
+    #         raise ValueError('Invalid base')
+    #     else:
+    #         base = int(base)
+
+    # Needed in the general case, but not as implemented
+    # if base <= 1:
+    #     raise ValueError('Invalid base')
+
+    # Needed in the general case, but not as implemented
+    # if num == 0:
+    #     numDigs = 1
+    # else:
+    numDigs = math.ceil(math.log(num, base))
+    if math.log(num, base).is_integer():
+        numDigs += 1
+
+    digs = [0.0 for i in range(0, numDigs)]
+    rem = num
+    for i in range(0, numDigs):
+        ix = numDigs - i - 1
+        dg = math.floor(rem / base**ix)
+        rem = rem % base**ix
+        digs[i] = dg
+    return digs
+
+
+def indexCorrector(ixs, base):
+    for i in range(0, len(ixs)):
+        ix = ixs[i]
+        if i + 1 < len(ixs):
+            if ixs[i + 1] == 0:
+                ixs[i] -= 1
+                ixs[i + 1] = base
+                if ixs[i] == 0:
+                    ixs = indexCorrector(ixs, base)
+    return ixs
+
+
+def alphabetStringGenerator(num, indexMode=False):
+    alphabet = ['.', 'i', 'j', 'k', 'm', 'n', 'p', 'q', 'r']
+
+    ixs = decoder(num + 1, len(alphabet) - 1)
+    pstr = ''
+    ixs = indexCorrector(ixs, len(alphabet) - 1)
+    for i in range(0, len(ixs)):
+        ix = ixs[i]
+        pstr += alphabet[ix]
+        pstr = pstr.replace('.', '')
+    return pstr
+
+
+def templatize_expression(expr):
+    expr, indices = templatize_rule(expr.parent_block(), expr._rule, expr.index_set())
+    return (expr, indices)
+
+
+def templatize_passthrough(con):
+    return (con, [])
+
+
+def precedenceChecker(node, arg1, arg2=None):
+    childPrecedence = []
+    for a in node.args:
+        if hasattr(a, 'PRECEDENCE'):
+            if a.PRECEDENCE is None:
+                childPrecedence.append(-1)
+            else:
+                childPrecedence.append(a.PRECEDENCE)
+        else:
+            childPrecedence.append(-1)
+
+    if hasattr(node, 'PRECEDENCE'):
+        precedence = node.PRECEDENCE
+    else:
+        # Should never hit this
+        raise DeveloperError(
+            'This error should never be thrown, node does not have a precedence.  Report to developers'
+        )
+
+    if childPrecedence[0] > precedence:
+        arg1 = ' \\left( ' + arg1 + ' \\right) '
+
+    if arg2 is not None:
+        if childPrecedence[1] > precedence:
+            arg2 = ' \\left( ' + arg2 + ' \\right) '
+
+    return arg1, arg2
+
+
+def handle_negation_node(visitor, node, arg1):
```

Comment:
For this and all that follow, the functions need to have the same first two inputs for the scheme to work because some of the functions do use visitor